Add example of unintentional None to 2.19 porting guide (#2982)
ansible/ansible-documentation: docs/docsite/rst/porting_guides/porting_guide_core_2.19.rst

    }

Example - unintentional ``None`` result
"""""""""""""""""""""""""""""""""""""""

If a template evaluated to ``None``, it was implicitly converted to an empty string in previous versions of ansible-core.

> This can now result in the template evaluating to the *value* ``None``.

Reviewer: It's worth pointing out that this no longer matters for the common case of passing the template result to a module argument that accepts a string, since `None` is now automatically converted to an empty string in that case.

See: https://github.com/ansible/ansible/pull/85652

Author: Ah, I didn't notice that change. That doesn't work for `type=list` though, where an empty string as a result of a template in 2.18 and before resulted in `[""]`, and now results in an error. But since `[""]` was seldom the expected value I guess this doesn't matter much...

Author: I added a note: dbe43f4f2e230bd7eebe0ae5be76785333fe2369
Does this look good?

Reviewer: This looks better.

Your point about lists has me wondering if they should default to `[]` when given `None` for module arguments, since that's the closest thing to "no list" -- just like we now do for strings defaulting to "" when `None` is given.

Author: Defaulting to `[]` would be changing behavior (since depending on where `None` comes from, it used to be `[""]` before). That sounds like a can of worm that can cause more bad than good. At least for strings, `None` either came from a template (and used to be `""`) or was invalid before.

The following example shows a case where this happens:

.. code-block:: yaml+jinja

    - set_fact:
        # If 'foo' is not defined, the else branch basically evaluates to None.
        # So value_none will not be an empty string, but None:
        value_none: |-
          {% if foo is defined %}foo is defined{% endif %}

This example can be fixed as follows:

.. code-block:: yaml+jinja

    - set_fact:
        # Explicitly return an empty string in the 'else' branch.
        # The value is always a string: either "foo is defined" or "".
        value_none: |-
          {% if foo is defined %}foo is defined{% else %}{{ "" }}{% endif %}

> This adjustment also work fine with older ansible-core versions.

Reviewer 2: ```suggestion
This adjustment is backward-compable with older ansible-core versions.
```

> Example - unintentional ``None`` preventing native output

Reviewer: This second example is no longer correct now that https://github.com/ansible/ansible/pull/85676 has been merged.

Author: I was pretty sure I re-tested it, but apparently I did something wrong, it indeed no longer happens. I'll cut that out.

Author: 163eb58316c9575bcb5b46290ecfaf8df22abe9f

The following example shows a case where this happens:

.. code-block:: yaml+jinja

    - set_fact:
        # The expression 'items.append(x)' evaluates to None, so the template evalutes to
        # three None's followed by the list ['a', 'b', 'c'].
        # In older ansible-core versions, the way evaluation was done this still resulted in a list.
        # Now, it results in a stringified version of the list instead:  "['a', 'b', 'c']".
        evaluate_to_string_instead_of_list: |-
          {% set items = [] %}
          {% for x in ['a', 'b', 'c'] %}
          {{-  items.append(x) -}}
          {% endfor %}
          {{ items }}

This example can be fixed as follows:

.. code-block:: yaml+jinja

    - set_fact:
        # Using {% set _ = expression %} instead of {{ expression }} eats the
        # return value.  The template evalutes to the array ['a', 'b', 'c']:
        evaluate_to_string_instead_of_list: |-
          {% set items = [] %}
          {% for x in ['a', 'b', 'c'] %}
          {%-  set _ = items.append(x) -%}
          {% endfor %}
          {{ items }}

> This adjustment also work fine with older ansible-core versions.

Reviewer 2: ```suggestion
This adjustment is backward-compable with older ansible-core versions.
```
